### PR TITLE
tackle-76-get-assessment

### PR DESCRIPTION
**Pull request by jonathanvila**
Issue : https://github.com/konveyor/tackle-pathfinder/issues/12

### Features covered
* New endpoint GET /assessments/{assessmentId} that will return the assessment with categories, questions, options, stakeholders, stakeholdergroups, comments and selection

### Pre steps : 
1. Install minikube : https://minikube.sigs.k8s.io/docs/start/
1. Start minikube : minikube start --kubernetes-version=v1.20.2 --cpus 4 --memory 12000
1. Enable ingress addons : minikube addons enable ingress
1. Create tackle namespace
```Shell
kubectl create namespace tackle
``` 
1. Deploy tackle ( to have the keycloak instance running ) on your K8s cluster
```Shell
$ kubectl apply -f  https://raw.githubusercontent.com/konveyor/tackle/main/kubernetes/kubernetes-tackle.yaml -n tackle
```
5. Edit the Ingress object to allow direct request to Pathfinder API
```Shell
$ kubectl edit ingress tackle -n tackle
```
Add this below `paths:`
```Yaml
     - backend:
          service:
            name: tackle-pathfinder
            port:
              number: 8080 
        path: /pathfinder
        pathType: ImplementationSpecific
```
6. Build and push 
```Shell
$ ./mvnw -U -B package -Dquarkus.container-image.push=true -Dquarkus.container-image.group={your quay user} -Dquarkus.container-image.registry=quay.io -Dquarkus.container-image.username={your quay user} -Dquarkus.container-image.password={your quay pwd} -Pnative
```
7. Change the image in the deployment tackle-pathfinder to be the one you have pushed
8. Rollout to update the image
```Shell
$ kubectl rollout restart deployment tackle-pathfinder -n tackle
```
### Envrioment vars for convenience with the cluster IP
```Shell
export api_ip={cluster ip}
export kc_ip={cluster ip}
```
### Test case 1 : Obtain Keycloak secret
```Shell
$ export access_token=$(\
    curl -X POST http://$kc_ip/auth/realms/quarkus/protocol/openid-connect/token \
    --user backend-service:secret \
    -H 'content-type: application/x-www-form-urlencoded' \
    -d 'username=alice&password=alice&grant_type=password' | jq --raw-output '.access_token' \
 )
```
> Expectation test case 1
```Shell
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100  1825  100  1776  100    49  18500    510 --:--:-- --:--:-- --:--:-- 19010
```

### Test case 2 : Create assessment for applicationId=20
```Shell
curl "http://$api_ip/pathfinder/assessments" \
  -H 'Content-Type: application/json' -H 'Accept: application/json' -H "Authorization: Bearer "$access_token \
  -d "{ \"applicationId\": 20 }" \
  -s -i
```
> Expectation Test case 2
```Shell
HTTP/1.1 201 Created
Date: Mon, 22 Mar 2021 11:05:29 GMT
Content-Type: application/json
Content-Length: 46
Connection: keep-alive

{"id": {any ID},"applicationId":20,"status":"STARTED"}
```

### Test case 3 : Get the created Assessment
```Shell
curl -X GET "http://$api_ip/pathfinder/assessments/{previous ID}" \
  -H 'Content-Type: application/json' -H 'Accept: application/json' -H "Authorization: Bearer $access_token"  -s -i
```

> Expectation Test Case 3
( something similar to this output )
```Shell
HTTP/1.1 200 OK
Date: Wed, 14 Apr 2021 12:23:23 GMT
Content-Type: application/json
Transfer-Encoding: chunked
Connection: keep-alive

{"id":418,"applicationId":92,"status":"STARTED","stakeholders":[],"stakeholderGroups":[],"questionnaire":{"categories":[{"id":420,"order":1,"title":"Application Details","questions":[{"id":421,"order":1,"question":"Does the application development team understand and actively develop the application?","options":[{"id":422,"order":0,"option":"Unknown","checked":false},{"id":423,"order":1,"option":"External ..........
```

### CODE REVIEW

File: src/main/java/io/tackle/pathfinder/services/AssessmentSvc.java [hunk 1]
```diff
@@ -56,33 +57,47 @@ public AssessmentHeaderDto createAssessment(@NotNull Long applicationId) {
...
 
         assessment.assessmentQuestionnaire.categories = new ArrayList<AssessmentCategory>();
         for (Category category : questionnaire.categories) {
-            AssessmentCategory assessmentCategory = AssessmentCategory.builder().name(category.name)
-                    .order(category.order).questionnaire(assessment.assessmentQuestionnaire )
-                    .questions(new ArrayList<>()).build();
+            AssessmentCategory assessmentCategory = AssessmentCategory.builder()
+                    .name(category.name)
+                    .order(category.order)
+                    .questionnaire(assessment.assessmentQuestionnaire )
```
Comment (m-brophy) [inline src/main/java/io/tackle/pathfinder/services/AssessmentSvc.java:77]: Why don't we instantiate the questions here any more?

File: src/main/java/io/tackle/pathfinder/services/AssessmentSvc.java [hunk 2]
```diff
@@ -56,33 +57,47 @@ public AssessmentHeaderDto createAssessment(@NotNull Long applicationId) {
...
             for (Question question : category.questions) {
-                AssessmentQuestion assessmentQuestion = AssessmentQuestion.builder().category(assessmentCategory)
-                        .name(question.name).order(question.order).questionText(question.questionText).type("SINGLE")
-                        .singleOptions(new ArrayList<>()).build();
+                AssessmentQuestion assessmentQuestion = AssessmentQuestion.builder()
+                        .category(assessmentCategory)
+                        .name(question.name)
+                        .order(question.order)
+                        .questionText(question.questionText)
+                        .type(question.type)
```
Comment (m-brophy) [inline src/main/java/io/tackle/pathfinder/services/AssessmentSvc.java:87]: Is there a closed set of possible question types? Is that set up elsewhere, maybe in another project?

File: src/main/resources/db/migration/V0.0.2_003__insert_default_questionnaire.sql [hunk 3]
```diff
@@ -3599,4 +3599,4 @@ values (
             select max(id)
             from question
         )
-    );
\ No newline at end of file
+    );
```
Comment (m-brophy) [inline src/main/resources/db/migration/V0.0.2_003__insert_default_questionnaire.sql:3602]: we don't need to change this file just for a new line at the end

File: src/main/resources/db/migration/V0.0.2_001__create_questionaire.sql [hunk 4]
```diff
@@ -54,4 +54,4 @@ add constraint FKjqvqo75p8dpn6oweuepac4ubf foreign key (questionnaire_id) refere
 alter table if exists question
 add constraint FK7jaqbm9p4prg7n91dd1uabrvj foreign key (category_id) references category;
 alter table if exists single_option
-add constraint FKi8whcw3jwcfm8sklck50a6g0e foreign key (question_id) references question;
\ No newline at end of file
+add constraint FKi8whcw3jwcfm8sklck50a6g0e foreign key (question_id) references question;
```
Comment (m-brophy) [inline src/main/resources/db/migration/V0.0.2_001__create_questionaire.sql:57]: we don't need to change this file just for a new line at the end

File: doc/diagrams/src/usecases.assess.plantuml [hunk 5]
```diff
@@ -11,6 +11,8 @@ package Copy {
     usecase "Copy Assessment" as copyfromto
 }
 package Assess {
+    usecase "Create new unique Assessment for Application" as newassessment
```
[the remark this answers is not in the document]
Comment (jonathanvila) [reply doc/diagrams/src/usecases.assess.plantuml:14]: :+1:

File: src/main/java/io/tackle/pathfinder/controllers/AssessmentsResource.java [hunk 6]
```diff
@@ -42,4 +43,11 @@ public Response createAssessment(@NotNull @Valid ApplicationDto data) {
       .build();
   }
 
+  @GET
+  @Path("{assessmentId}")
+  @Produces("application/json")
+  public Response getAssessment(@NotNull @PathParam("assessmentId") Long assessmentId) {
```
[the remark this answers is not in the document]
Comment (jonathanvila) [reply src/main/java/io/tackle/pathfinder/controllers/AssessmentsResource.java:49]: :+1:

File: src/main/java/io/tackle/pathfinder/services/AssessmentSvc.java (hunk 1 above)
Comment (jonathanvila) [reply src/main/java/io/tackle/pathfinder/services/AssessmentSvc.java:77]: Is it already instantiated in the Entity

File: src/main/java/io/tackle/pathfinder/services/AssessmentSvc.java (hunk 2 above)
Comment (jonathanvila) [reply src/main/java/io/tackle/pathfinder/services/AssessmentSvc.java:87]: Good spot. 
Converting it to an Enum of SINGLE at the moment

File: src/main/resources/db/migration/V0.0.2_001__create_questionaire.sql (hunk 4 above)
Comment (jonathanvila) [reply src/main/resources/db/migration/V0.0.2_001__create_questionaire.sql:57]: I simply added EOF character

File: src/main/resources/db/migration/V0.0.2_003__insert_default_questionnaire.sql (hunk 3 above)
Comment (jonathanvila) [reply src/main/resources/db/migration/V0.0.2_003__insert_default_questionnaire.sql:3602]: I simply added EOF character

**Review by m-brophy — APPROVED** [review]

**Pull request merged**